#**Telecom Customer Churn Prediction**

***Supervised ML model: Prediction acccuracy needs to be High here***

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

***Download the Datasets***

In [3]:
df= pd.read_csv('https://github.com/YBIFoundation/Dataset/raw/main/TelecomCustomerChurn.csv')

In [4]:
df.head()

customerID  Gender  SeniorCitizen Partner Dependents  Tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

  MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0            No             DSL             No  ...               No   
1            No             DSL            Yes  ...              Yes   
2            No             DSL            Yes  ...               No   
3            No             DSL            Yes  ...              Yes   
4            No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies  Contract PaperlessBilling  \
0          No          No              No   Monthly              Yes   
1          No          No              No  One year               No   
2          No          No              No   Monthly              Yes   
3         Yes          No              No  One year               No   
4          No          No              No   Monthly              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0                     Manual          29.85         29.85    No  
1                     Manual          56.95        1889.5    No  
2                     Manual          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4                     Manual          70.70        151.65   Yes  

[5 rows x 21 columns]

#****Exploratory Data Analysis****

In [5]:
df.describe()

SeniorCitizen       Tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   Tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [7]:
df.columns

Index(['customerID', 'Gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'Tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [8]:
df.duplicated().sum()

0

***This dataset dont have a duplicate***

***Define y & X***

In [9]:
y= df['Churn']
X= df.drop(['customerID','Churn'], axis=1)

In [10]:
X

Gender  SeniorCitizen Partner Dependents  Tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

     MultipleLines InternetService OnlineSecurity OnlineBackup  \
0               No             DSL             No          Yes   
1               No             DSL            Yes           No   
2               No             DSL            Yes          Yes   
3               No             DSL            Yes           No   
4               No     Fiber optic             No           No   
...            ...             ...            ...          ...   
7038           Yes             DSL            Yes           No   
7039           Yes     Fiber optic             No          Yes   
7040            No             DSL            Yes           No   
7041           Yes     Fiber optic             No           No   
7042            No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies  Contract  \
0                  No          No          No              No   Monthly   
1                 Yes          No          No              No  One year   
2                  No          No          No              No   Monthly   
3                 Yes         Yes          No              No  One year   
4                  No          No          No              No   Monthly   
...               ...         ...         ...             ...       ...   
7038              Yes         Yes         Yes             Yes  One year   
7039              Yes          No         Yes             Yes  One year   
7040               No          No          No              No   Monthly   
7041               No          No          No              No   Monthly   
7042              Yes         Yes         Yes             Yes  Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0                 Yes                     Manual           29.85        29.85  
1                  No                     Manual           56.95       1889.5  
2                 Yes                     Manual           53.85       108.15  
3                  No  Bank transfer (automatic)           42.30      1840.75  
4                 Yes                     Manual           70.70       151.65  
...               ...                        ...             ...          ...  
7038              Yes                     Manual           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes                     Manual           29.60       346.45  
7041              Yes                     Manual           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5  

[7043 rows x 19 columns]

***We have to encode the categorical data***

***One-Hot Encoding***

In [12]:
from sklearn.preprocessing import OrdinalEncoder
oe= OrdinalEncoder()

In [13]:
 X=oe.fit_transform(X)

In [14]:
X

array([[0.000e+00, 0.000e+00, 1.000e+00, ..., 2.000e+00, 1.420e+02,
        2.505e+03],
       [1.000e+00, 0.000e+00, 0.000e+00, ..., 2.000e+00, 4.980e+02,
        1.466e+03],
       [1.000e+00, 0.000e+00, 0.000e+00, ..., 2.000e+00, 4.360e+02,
        1.570e+02],
       ...,
       [0.000e+00, 0.000e+00, 1.000e+00, ..., 2.000e+00, 1.370e+02,
        2.994e+03],
       [1.000e+00, 1.000e+00, 1.000e+00, ..., 2.000e+00, 7.950e+02,
        2.660e+03],
       [1.000e+00, 0.000e+00, 0.000e+00, ..., 0.000e+00, 1.388e+03,
        5.407e+03]])

***Scaling***

In [15]:
y.value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

***Here we have to convert Imbalance Dataset to balanced***

In [16]:
from imblearn.over_sampling import RandomOverSampler
ros= RandomOverSampler()

In [17]:
X,y= ros.fit_resample(X,y)

In [18]:
y.value_counts()

No     5174
Yes    5174
Name: Churn, dtype: int64

#***Train Test Split***

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(X, y, random_state=2529)

***Select the model: Ensemble(Random Forest)***

In [21]:
from sklearn.ensemble import RandomForestClassifier
rfc= RandomForestClassifier()

***Train the Model***

In [22]:
rfc.fit(X_train, y_train)

RandomForestClassifier()

***Model Prediction***

In [24]:
y_pred= rfc.predict(X_test)

***Model Evaluation***

In [25]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          No       0.95      0.81      0.87      1290
         Yes       0.83      0.96      0.89      1297

    accuracy                           0.88      2587
   macro avg       0.89      0.88      0.88      2587
weighted avg       0.89      0.88      0.88      2587



***Recall is very good for this Model stands AT 96% which says that this model is able to catch majority od customers***